File: apps/web/lib/providers/crypto.ts
import { createCipheriv, createDecipheriv, randomBytes, createHash } from "crypto";

/**
 * Application-level envelope encryption for `tenant_provider_config.config`
 * (and, in Phase 3, any other per-tenant provider secret column). See
 * docs/PROVIDER_REGISTRY.md "Secrets" — the DB never stores plaintext
 * credentials; encryption/decryption happens here, in the app process,
 * keyed by PROVIDER_CONFIG_ENCRYPTION_KEY (an env var, never in the DB).
 *
 * TODO(infra, before real tenant secrets go live): replace this
 * single-symmetric-key-from-env approach with a real secrets manager/KMS
 * (per-tenant data keys, rotation, audit log of decrypt calls). This is
 * intentionally the minimum viable "never plaintext in Postgres" bar for
 * Phase 2, not the final design.
 */

const ALGORITHM = "aes-256-gcm";

export type EncryptedConfig = {
  iv: string;
  tag: string;
  ciphertext: string;
};

function getKey(): Buffer {
  const raw = process.env.PROVIDER_CONFIG_ENCRYPTION_KEY;
  if (!raw) {
    throw new Error(
      "PROVIDER_CONFIG_ENCRYPTION_KEY is not set. Required to encrypt/decrypt tenant_provider_config.config. " +
        "See .env.example."
    );
  }
  // Accept any-length passphrase for dev ergonomics; derive a stable 32-byte
  // key via SHA-256 rather than requiring an exact 32-byte hex/base64 value.
  return createHash("sha256").update(raw, "utf8").digest();
}

export function encryptProviderConfig(plaintext: Record<string, unknown>): EncryptedConfig {
  const key = getKey();
  const iv = randomBytes(12);
  const cipher = createCipheriv(ALGORITHM, key, iv);
  const json = Buffer.from(JSON.stringify(plaintext), "utf8");
  const ciphertext = Buffer.concat([cipher.update(json), cipher.final()]);
  const tag = cipher.getAuthTag();
  return {
    iv: iv.toString("base64"),
    tag: tag.toString("base64"),
    ciphertext: ciphertext.toString("base64"),
  };
}

export function decryptProviderConfig(encrypted: EncryptedConfig): Record<string, unknown> {
  const key = getKey();
  const decipher = createDecipheriv(ALGORITHM, key, Buffer.from(encrypted.iv, "base64"));
  decipher.setAuthTag(Buffer.from(encrypted.tag, "base64"));
  const plaintext = Buffer.concat([
    decipher.update(Buffer.from(encrypted.ciphertext, "base64")),
    decipher.final(),
  ]);
  return JSON.parse(plaintext.toString("utf8"));
}

export function isEncryptedConfig(value: unknown): value is EncryptedConfig {
  return (
    typeof value === "object" &&
    value !== null &&
    typeof (value as EncryptedConfig).iv === "string" &&
    typeof (value as EncryptedConfig).tag === "string" &&
    typeof (value as EncryptedConfig).ciphertext === "string"
  );
}

File: services/voice-gateway/voice_gateway/crypto.py
"""Envelope decryption for `tenant_provider_config.config`.

Mirrors apps/web/lib/providers/crypto.ts byte-for-byte (same algorithm, same
key derivation, same envelope shape) so a config row encrypted by the
Next.js app can be decrypted here, and vice versa — this is the one place
the two runtimes must agree on a wire format, since they share the same
Postgres row. See docs/PROVIDER_REGISTRY.md "Secrets".

Envelope shape (JSON, as stored in the `config` jsonb column):
    {"iv": "<base64, 12 bytes>", "tag": "<base64, 16 bytes>", "ciphertext": "<base64>"}

Algorithm: AES-256-GCM. Key: SHA-256 of the UTF-8 PROVIDER_CONFIG_ENCRYPTION_KEY
env var (accepts any-length passphrase for dev ergonomics, exactly like the
TS side's `createHash("sha256").update(raw, "utf8").digest()`).

TODO(infra, same note as crypto.ts): single symmetric key from an env var is
the Phase 2/3 minimum bar, not the final design — replace with a real
secrets manager/KMS before onboarding real tenant secrets.
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
from typing import Any

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


class EncryptionKeyMissingError(RuntimeError):
    pass


def _get_key() -> bytes:
    raw = os.environ.get("PROVIDER_CONFIG_ENCRYPTION_KEY")
    if not raw:
        raise EncryptionKeyMissingError(
            "PROVIDER_CONFIG_ENCRYPTION_KEY is not set. Required to decrypt "
            "tenant_provider_config.config. See .env.example."
        )
    return hashlib.sha256(raw.encode("utf-8")).digest()


def is_encrypted_config(value: Any) -> bool:
    return (
        isinstance(value, dict)
        and isinstance(value.get("iv"), str)
        and isinstance(value.get("tag"), str)
        and isinstance(value.get("ciphertext"), str)
    )


def decrypt_provider_config(encrypted: dict[str, str]) -> dict[str, Any]:
    key = _get_key()
    iv = base64.b64decode(encrypted["iv"])
    tag = base64.b64decode(encrypted["tag"])
    ciphertext = base64.b64decode(encrypted["ciphertext"])
    aesgcm = AESGCM(key)
    # AESGCM expects ciphertext+tag concatenated (cryptography's API differs
    # from Node's split ciphertext/authTag — reassemble to match).
    plaintext = aesgcm.decrypt(iv, ciphertext + tag, None)
    return json.loads(plaintext.decode("utf-8"))


def encrypt_provider_config(plaintext: dict[str, Any]) -> dict[str, str]:
    """Provided for symmetry/tests; the Python side is a reader in
    production (tenant config is written via apps/web's admin tooling), but
    tests need to produce a valid envelope without a live Node process."""
    key = _get_key()
    iv = os.urandom(12)
    aesgcm = AESGCM(key)
    raw = aesgcm.encrypt(iv, json.dumps(plaintext).encode("utf-8"), None)
    ciphertext, tag = raw[:-16], raw[-16:]
    return {
        "iv": base64.b64encode(iv).decode(),
        "tag": base64.b64encode(tag).decode(),
        "ciphertext": base64.b64encode(ciphertext).decode(),
    }
